src/core/auth/wallet-authenticator.ts

```diff
@@ -0,0 +1,189 @@
+import { ethers } from "ethers";
+import { SDKOptions } from "../../schema";
+import {
+  AuthenticatedPayload,
+  AuthenticatedPayloadSchema,
+  AuthenticationPayloadInput,
+  AuthenticationPayloadInputSchema,
+  AuthenticationPayloadOutput,
+  AuthenticationPayloadOutputSchema,
+  AuthorizedPayload,
+  AuthorizedPayloadSchema,
+  FilledAuthenticationPayloadInput,
+} from "../../schema/auth";
+import { RPCConnectionHandler } from "../classes/rpc-connection-handler";
+import { NetworkOrSignerOrProvider } from "../types";
+
+/**
+ * Wallet Authenticator
+ * @remarks Enables sign-in with ethereum based authentication.
+ * @example
+ * ```javascript
+ * // We choose an application name on the server side
+ * const application = "my-app-name"
+ *
+ * // On the server side, we can generate a payload for a wallet requesting to authenticate
+ * const payloadWithSignature = await sdk.auth.generate({
+ *   application,
+ *   subject: "0x...",
+ * })
+ *
+ * // Then on the client side, we can sign this payload with the wallet requesting to authenticate
+ * const authenticatedPayload = await sdk.auth.sign(payloadWithSignature)
+ *
+ * // Finally, on the server side, we can verify if this payload is valid
+ * const isValid = await sdk.auth.verify(application, authenticatedPayload)
+ * ```
+ * @public
```

Comment:
let's mark this @beta for now. We need to make sure this is rock solid with no possible exploit before we go full prod